doc/docusaurus/docs/essential-concepts/versions.md

```diff
@@ -0,0 +1,68 @@
+---
+sidebar_position: 5
+---
+
+# Different Notions of Version
+
+There are several different notions of version that Cardano smart contract developers must distinguish.
+
+## Ledger Language Version
+
+This is what "Plutus V1", "Plutus V2", "Plutus V3" refer to.
+It is called "ledger" language version because they are _not_ really different versions of a language in the sense of programming languages, but different languages from the Cardano ledger's point of view; the ledger handles Plutus V1 vs. V2 vs. V3 differently.
+In essence, "Plutus V1", "Plutus V2" and "Plutus V3" are tags that the ledger attaches to validators in a transaction.
+
+From the ledger's perspective, they are in fact totally different languages:
+there is no requirement that they be similar or compatible in any way.
+However, the "V1", "V2" and "V3" naming scheme is practically useful because in reality they have a lot in common (e.g., the underlying language is Plutus Core and programs are executed by the Plutus Core evaluator, regardless of ledger language version); plus it makes it clear the order that they were introduced and the relationships among them.
+
+When simply writing a standalone program, say one that takes two integers and returns their sum, or one that implements a Sudoku solver, the notion of ledger language version is completely irrelevant.
+There is no need (and nowhere) to specify whether you are writing a Plutus V1, or Plutus V2, or Plutus V3 program - again, the concept of ledger language version does not apply to programs themselves.
+You simply write your code in Plutus Tx, compile it to UPLC, and run it with a UPLC evaluator.
```

Comment:
I believe I understand what this is trying to say but I also think it's a bit ambiguous. It relies on "language in the sense of a programming language" having a clear definition when one is not given here and it's also not common knowledge what it might be.

How can we define what a "language version" is?
Would it make sense to say that the versioning `PlutusV1 ... PlutusVn` refers to the version of the UPLC evaluator? I believe that's not enough, because if `V2` added new constructors to the UPLC AST then it also extended the syntax.
So is it a new UPLC language? Where the syntax can only be extended from `Vn` to `Vn+1` and the semantics can be changed arbitrarily? Or can only the "builtin" semantics be changed arbitrarily? It's not clear to me what constraints we have here, and it would be very good to specify them.

All of these are independent of the ledger, so I'm not so sure of the "_ledger_ language version" naming. I would say that yes the ledger interacts with each language version differently, but the "languages" themselves are different in isolation of the ledger. And I believe we should (or, more appropriately, the ledger should) specify how it "consumes" each language version, since from "Depending on the ledger language version, the ledger will pass different arguments to the validator." I understand that the API differs from one version to the other.

Reply:
> Would it make sense to say that the versioning PlutusV1 ... PlutusVn refers to the version of the UPLC evaluator?

No, PlutusVn has nothing whatsoever to do with the UPLC evaluator.

> but the "languages" themselves are different in isolation of the ledger.

No, the languages themselves are not different, which is the entire point I'm trying to make here. How are they different in isolation of the ledger?

Reply:
> No, the languages themselves are not different, which is the entire point I'm trying to make here. How are they different in isolation of the ledger?

`PlutusV2` introduces new constructors to the UPLC language, right? What does this have to do with the ledger?

Reply:
> PlutusV2 introduces new constructors to the UPLC language, right? What does this have to do with the ledger?

No it doesn't. New constructors were introduced in a new Plutus Core Version (1.1.0), not in PlutusV2.

Reply:
> The ledger language version (e.g. PlutusV1) literally determines the AST version (1.0.0 for PlutusV1)

Again, this will no longer be true in the near future. There is no fundamental reason why 1.1.0 cannot be used for PlutusV1.